### PR TITLE
[lld-macho] Implement ObjC category merging (-objc_category_merging)

### PR DESCRIPTION
**Pull request by alx32**
This change adds a flag to lld to enable category merging for MachoO + ObjC.
It adds the '-objc_category_merging' flag for enabling this option and uses the existing '-no_objc_category_merging' flag for disabling it.
In ld64, this optimization is enabled by default, but in lld, for now, we require explicitly passing the '-objc_category_merging' flag in order to enable it.

Behavior: if in the same link unit, multiple categories are extending the same class, then they get merged into a single category.
Ex: Cat1(method1+method2,protocol1) + Cat2(method3+method4,protocol2, property1) = Cat1_2(method1+method2+method3+method4, protocol1+protocol2, property1)

Notes on implementation decisions made in this diff:

There is a possibility to further improve the current implementation by directly merging the category data into the base class (if the base class is present in the link unit) - this improvement may be done as a follow-up. This improved functionality is already present in ld64.
We do the merging on the raw inputSections - after dead-stripping (categories can't be dead stripped anyway).
The changes are mostly self-contained to ObjC.cpp, except for adding a new flag (linkerOptimizeReason) to ConcatInputSection and StringPiece to mark that this data has been optimized away. Another way to do it would have been to just mark the pieces as not 'live' but this would cause the old symbols to show up in the linker map as being dead-stripped - even if dead-stripping is disabled. This flag allows us to match the ld64 behavior.

Note: This is a re-land of https://github.com/llvm/llvm-project/pull/82928 after fixing using already freed memory in `generatedSectionData`.  This issue was detected by ASAN build.

### CODE REVIEW

**Comment by alx32** [comment]
Note: This is a re-land of https://github.com/llvm/llvm-project/pull/82928 after fixing using already freed memory in generatedSectionData. This issue was detected by ASAN build.

File: lld/MachO/ObjC.cpp [hunk 1]
```diff
@@ -320,3 +332,908 @@ void objc::checkCategories() {
...
+
+  assert(!ptrList.structSize || (thisStructSize == ptrList.structSize));
+
+  ptrList.structCount += thisStructCount;
+  ptrList.structSize = thisStructSize;
+
+  uint32_t expectedListSize =
+      listHeaderLayout.totalSize + (thisStructSize * thisStructCount);
+  assert(expectedListSize == ptrListSym->isec->data.size() &&
+         "Pointer list does not match expected size");
```
Comment (kyulee-com) [inline lld/MachO/ObjC.cpp:691]: I think you need to add the following, to ensure this release build without assertion like https://github.com/llvm/llvm-project/commit/6ae77eca6ceb17033eb58440038e620eba1c6f51.
```
(void)expectedListSize;
```

File: lld/MachO/ObjC.cpp [hunk 2]
```diff
@@ -320,3 +332,908 @@ void objc::checkCategories() {
...
+  assert(categories.size() > 1 && "Expected at least 2 categories");
+
+  ClassExtensionInfo extInfo(catLayout);
+
+  for (auto &catInfo : categories)
+    parseCatInfoToExtInfo(catInfo, extInfo);
+
+  Defined *newCatDef = emitCategory(extInfo);
+  assert(newCatDef && "Failed to create a new category");
+
```
Comment (kyulee-com) [inline lld/MachO/ObjC.cpp:1038]: The same thing here.
```
(void)newCatDef;
```

File: lld/MachO/ObjC.cpp [hunk 3]
```diff
@@ -320,3 +332,914 @@ void objc::checkCategories() {
...
+  eraseMergedCategories();
+}
+
+void ObjcCategoryMerger::doCleanup() { generatedSectionData.clear(); }
+
+StringRef ObjcCategoryMerger::newStringData(const char *str) {
+  uint32_t len = strlen(str);
+  auto &data = newSectionData(len + 1);
+  char *strData = reinterpret_cast<char *>(data.data());
+  strncpy(strData, str, len);
```
Comment (mikaelholmen) [inline lld/MachO/ObjC.cpp:1226]: Hi @alx32

gcc warns on this code like
```
../../lld/MachO/ObjC.cpp:1226:10: warning: 'char* strncpy(char*, const char*, size_t)' output truncated before terminating nul copying as many bytes from a string as its length [-Wstringop-truncation]
 1226 |   strncpy(strData, str, len);
      |   ~~~~~~~^~~~~~~~~~~~~~~~~~~
../../lld/MachO/ObjC.cpp: In member function 'void {anonymous}::ObjcCategoryMerger::emitAndLinkPointerList(lld::macho::Defined*, uint32_t, const {anonymous}::ObjcCategoryMerger::ClassExtensionInfo&, const {anonymous}::ObjcCategoryMerger::PointerListInfo&)':
../../lld/MachO/ObjC.cpp:1223:24: note: length computed here
 1223 |   uint32_t len = strlen(str);
      |                  ~~~~~~^~~~~
```
Anything to worry about?

Comment (alx32) [reply lld/MachO/ObjC.cpp:1226]: Thanks for flagging. Not a bug, as the buffer (len+1) is guaranteed to be null-initialized, so the null terminator will be there, but we should fix the warning: https://github.com/llvm/llvm-project/pull/86091

**Pull request merged**